### DISCUSSION TITLE
Problem with ILI9341 display

### DISCUSSION THREAD

**Discussion by adr-bed** [arendst/Tasmota#16114] [Q&A] [answered]
Hello!
I've configured a D1 MINI board with ILI9341 display using wiring like this:

![ILI9341](https://user-images.githubusercontent.com/91057963/181795118-08770c65-9111-4181-8147-714d111163af.png)

Tasmota seems to recognize the display just fine, but the problem is that display shows some artifacts/glitches. Result for

`DisplayText This is a very long sample text`

is

![ILI9341_2](https://user-images.githubusercontent.com/91057963/181797687-6c011727-efa6-4cf8-8aa7-e534ff66ff59.png)

My display settings:
`{"Display":"Model":4,"Type":1,"Width":320,"Height":240,"Mode":0,"Dimmer":100,"Size":2,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}`

I have check another, exactly the same display, and Tasmota every time shows exactly the same result. The "grainy" rectangle in the middle of the screen moves around when I display different text. Does anyone have any suggestions?

**Comment by gemu2015** [comment, marked as answer]
from your photo it looks like your wires are far to long, they should be as short as possible they carry at least 20 Mhz signals

**Comment by SN-8023** [comment]
Was it the cables that were too long?

because I have the same problem. but I have my doubts about the theory with the too long cables.
- the interference is absolutely stable and it starts after exactly 240px
- my wires are just 10cm long, (i shortened them to 5cm, same error)
- in TextMode 0 the display works as expected (apart from the interference on the right side)
- in displaymode 1 the content incl. interference shifts a few pixels to the left with every new line output. *2

i have no idea, but to me it looks like the buffer is only 240px long with meaningful data, but 320px are sent. but the 80 missing pixels contain only garbage, or are not sent at all

```
CMD: display
MQT: stat/tasmota_9040C8/RESULT = {"Display":{"Model":4,"Type":1,"Width":320,"Height":240,"Mode":2,"Dimmer":50,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}
```

P.S. when turning on the whole display contains the same kind of pattern for a very short time, but only the 240 pixels turn black then
this pattern remains when DC pin is disconnected *3
- what does "DC" do?

very mysterious

*1 ![IMG_2243](https://github.com/arendst/Tasmota/assets/65110735/69695094-81c2-4184-a5ef-3ff7c2ced5d9)
*2 ![IMG_2244](https://github.com/arendst/Tasmota/assets/65110735/c2012e15-d46e-4d6b-ad26-428e404104c9)
*3 ![IMG_2245](https://github.com/arendst/Tasmota/assets/65110735/fb3385c5-1454-43d6-b588-399f7badfdb8)

Any ideas or clues?

**Reply by sfromis** [reply]
I'd experiment with `DisplayRotate 1` (or 3), as your cutoff point of 240 pixels looks like your DisplayHeight of 240 is taken in the wrong orientation. The meaning of "DC" is data/command, which depending on the specific device type can be very important for the data flow. I'd expect nothing good from disconnecting it.

**Reply by SN-8023** [reply]
I had the same thought with DisplayRotate, but it is not impressed by my efforts. 
![IMG_2247](https://github.com/arendst/Tasmota/assets/65110735/69574e37-d893-4984-b938-2f7f1a0611e4)

i have also changed x and y in the ili9341_display.ini but i am not sure if i have messed around in the right place, because nothing has changed

I have also tried a second display, the same result.

**Reply by sfromis** [reply]
To use the `.ini` file, it should be renamed as `display.ini`, and you should also config some unused gpio as `Option A 3` to enable the [Universal Display Driver](https://tasmota.github.io/docs/Displays/#universal-display-driver), otherwise the older less flexible display driver will be used, as indicated by DisplayModel 4 instead of 17. And then you gets lots of possible adjustments there.

**Reply by SN-8023** [reply]
What I also noticed is that the colors are not correct.

`{"DisplayText":"[f1B0C31s1l15c1]1234567890abcdef1234567890abcdef"}
`
C31 should give a blue text and not a reddish one, or am I wrong?

regarding the display.ini, i'm afraid i don't really know which screws i should turn there.

**Reply by sfromis** [reply]
My point about `display.ini` is that this is the more flexible way, and preferable even before you consider where to "screw around". I'm always using the universal display driver for ILI9341, with the standard ini file. Were you picking up this one?
https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ILI9341_display.ini
But it will only come into effect with `Option A 3` on some unused gpio.

And I have zero trouble, including getting blue with your test example. Maybe get the pixel salad working before focusing on colors.

And make sure to use the full wiring for SPI, including DC. Here's pins from a working setup of mine:
```
Gpio.14  IO    768  SPI CS1
Gpio.18  IO    736  SPI CLK1
Gpio.19  IO    672  SPI MISO1
Gpio.23  IO    704  SPI MOSI1
Gpio.27  IO    800  SPI DC1
Gpio.32  AO    992  Backlight
Gpio.33  AO   1024  Display Rst
Gpio.39  IA   6210  Option A3
```
Of course, the specific gpio numbers are simply to match my hardware setup; not something preferable.

**Reply by SN-8023** [reply]
I have followed the hardware SPI of the ESP32
so 
```
GPIO02 ILI9341 CS   -> CS
GPIO04 ILI9341 DC   -> DC
GPIO17 option A3
GPIO19 display RST  ->RST
GPIO21 SPI MOSI     ->SDA
GPIO22 SPI CLK      ->SDK
```
the display does not have more pins
![IMG_2253](https://github.com/arendst/Tasmota/assets/65110735/fd20b460-3820-4178-83db-351da33fe036)

what is also a little irritating is the seller's indication Viewing angle 12:00'clock and 240x320 (not 320x240)
![image](https://github.com/arendst/Tasmota/assets/65110735/37e08372-3f64-457c-8e1c-961fd2b9e2a2)

`21:39:41.780 CMD: display
21:39:41.788 MQT: stat/tasmota_9040C8/RESULT = {"Display":{"Model":17,"Type":1,"Width":320,"Height":240,"Mode":0,"Dimmer":50,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}`

basically it seems to work, only the alignment is just wrong and I do not know where I can fix that. I've been using Tasmota for a while, but I've only had to deal with the issue of custom firmware for 3 days...

i also tested the content of `ILI9341_M5Stack_display.ini` in the `display.ini` (which is 240x320), no success

**Reply by sfromis** [reply]
For use with the universal display driver, you should not select the ILI9341-specific SPI pins, but the general ones like I did. With a test without having a MISO pin configured, my display still worked.

And no, you have not followed the hardware SPI of ESP32, as the chip has hardware SPI on the pins you choose. You may have followed default, but that is not relevant when you do a Tasmota config. Of course if the pins you choose happens to be the same as the default, that's fine too, but those pins are just not special.

Did you use the display.ini I linked?

**Reply by SN-8023** [reply]
ah, ok...
yes, i just tried again with the content of the ILI9341_display.ini. Result as before.

i will try other pins tomorrow. Although I don't think it's because of the wrong pins, because the communication works basically. Only 240 instead of 320 pixels arrive. Either this is a property of the protocol that is not covered by the driver, or the driver is configured incorrectly for MY type of display.

Anyway, thank you for your patience with me.

Just to be sure. I'll flash this then with

![image](https://github.com/arendst/Tasmota/assets/65110735/b59ad25e-f49a-466b-8f9b-6d7bc12bbc3b)

, Right?

**Reply by SN-8023** [reply]
unfortunately nothing has changed with a different pin assignment 
![image](https://github.com/arendst/Tasmota/assets/65110735/bb58f2b6-d4dc-4038-aecc-b28690759550)

**Reply by sfromis** [reply]
Did you try to do as I suggested, instead of moving same pins around?

In general, I'd also stay away from gpio 16/17 unless confirmed to work, as these are probed for PSRAM during boot. They mostly work anyway, but I like to cut down on "interference".

**Reply by SN-8023** [reply]
in the meantime I had a little success,... after quite a lot of trial and error, I am now with SSPI and a display.ini on the ESP.
I have in the display.ini in the filesyetem reduced the xValue to 160 to see what happens. The display was only initialized to half, the rest was not reset - ok, that was to be expected.
Then I set the value back to 320 and restarted the ESP. And lo and behold, all of a sudden it initialized the whole display and it works as it should (except for the colors)...

and I don't know why....

now i'll have a look at the stuff with the colors.
![IMG_2256](https://github.com/arendst/Tasmota/assets/65110735/80aa31fc-3a2c-4b03-93bf-e5c7af03c4e5)

**Reply by SN-8023** [reply]
I have it now so far that it is usable for me
With color values from the Common colors table the green is not correct.
But the color indices are correct. That is enough for me.

thanks @sfromis 

```
:H,GMT028-05_display.ini,320,240,16,SPI,3,*,*,*,*,*,*,*,40
:S,2,1,1,0,40,20
:I
EF,3,03,80,02
CF,3,00,C1,30
ED,4,64,03,12,81
E8,3,85,00,78
CB,5,39,2C,00,34,02
F7,1,20
EA,2,00,00
C0,1,23
C1,1,10
C5,2,3e,28
C7,1,86
36,1,48
37,1,00
3A,1,55
B1,2,00,18
B6,3,08,82,27
F2,1,00
26,1,01
E0,0F,0F,31,2B,0C,0E,08,4E,F1,37,07,10,03,0E,09,00
E1,0F,00,0E,14,03,11,07,31,C1,48,08,0F,0C,31,36,0F
11,80
29,80
:o,28
:O,29
:A,2A,2B,2C
:R,36
:0,40,00,00,00
:1,28,00,00,01
:2,88,00,00,02
:3,E8,00,00,03
#
```
